include/irr/asset/IVirtualTexture.h

```diff
@@ -0,0 +1,750 @@
+        assert(_pgTabLayers<=256u);
+
+        const uint32_t pgTabSzxy = 1u<<_pgTabSzxy_log2;
+        image_t::SCreationParams params;
+        params.arrayLayers = _pgTabLayers;
+        params.extent = {pgTabSzxy,pgTabSzxy,1u};
+        params.format = EF_R16G16_UINT;
+        params.mipLevels = std::max(static_cast<int32_t>(_maxAllocatableTexSz_log2-_pgSzxy_log2+1u), 1);
+        params.samples = IImage::ESCF_1_BIT;
+        params.type = IImage::ET_2D;
+        params.flags = static_cast<IImage::E_CREATE_FLAGS>(0);
+
+        auto pgtab = createImage(std::move(params));
+        IRR_PSEUDO_IF_CONSTEXPR_BEGIN(std::is_same<image_t,ICPUImage>::value)
+        {
+            const uint32_t texelSz = getTexelOrBlockBytesize(pgtab->getCreationParameters().format);
+        
+            auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(pgtab->getCreationParameters().mipLevels);
+
+            uint32_t bufOffset = 0u;
+            for (uint32_t i = 0u; i < pgtab->getCreationParameters().mipLevels; ++i)
+            {
+                const uint32_t tilesPerLodDim = pgTabSzxy>>i;
+                const uint32_t regionSz = _pgTabLayers*tilesPerLodDim*tilesPerLodDim*texelSz;
+                auto& region = (*regions)[i];
+                region.bufferOffset = bufOffset;
+                region.bufferImageHeight = 0u;
+                region.bufferRowLength = tilesPerLodDim;
+                region.imageExtent = {tilesPerLodDim,tilesPerLodDim,1u};
+                region.imageOffset = {0,0,0};
+                region.imageSubresource.baseArrayLayer = 0u;
+                region.imageSubresource.layerCount = _pgTabLayers;
+                region.imageSubresource.mipLevel = i;
+                region.imageSubresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+
+                bufOffset += regionSz;
+            }
+            auto buf = core::make_smart_refctd_ptr<ICPUBuffer>(bufOffset);
+            uint32_t* bufptr = reinterpret_cast<uint32_t*>(buf->getPointer());
+            std::fill(bufptr, bufptr+bufOffset/sizeof(uint32_t), SPhysPgOffset::invalid_addr);
+
+            pgtab->setBufferAndRegions(std::move(buf), regions);
+        } IRR_PSEUDO_IF_CONSTEXPR_END
+        return pgtab;
+    }
+
+    void updateLayerToViewIndexLUT(uint32_t _layer, uint32_t _smplrIndex)
+    {
+        (*m_layerToViewIndexMapping)[_layer] = _smplrIndex;
+        m_layer2viewWasUpdatedSinceLastQuery = true;
+    }
+
+    const uint32_t m_pgSzxy;
+    const uint32_t m_pgSzxy_log2;
+    const uint32_t m_pgtabSzxy_log2;
+    const uint32_t m_tilePadding;
+    core::smart_refctd_ptr<image_t> m_pageTable;
+
+    using pg_tab_addr_alctr_t = core::GeneralpurposeAddressAllocator<uint32_t>;
+    core::smart_refctd_dynamic_array<pg_tab_addr_alctr_t> m_pageTableLayerAllocators;
+    uint8_t* m_pgTabAddrAlctr_reservedSpc = nullptr;
+
+    core::stack<uint32_t> m_freePageTableLayerIDs;
```

Comment:
we have the `m_layerToViewIndexMapping` its better to `std:find_if(,~0u)` linear search from the front rather than `m_freePageTableLayerIDs` because then we'll always get the lowest numbered layer as the new layer